File: src/lib/rate-limit.ts
/**
 * Rate limiter sliding window in-memory.
 *
 * Pakai untuk endpoint yang memanggil API berbayar (chatbot, essay grading).
 * In-memory cocok untuk single-server deployment; untuk multi-instance,
 * ganti dengan Redis.
 */
type Bucket = { times: number[] };
const STORE_KEY = "__kwgn_rate_buckets__";
const g = globalThis as unknown as { [STORE_KEY]?: Map<string, Bucket> };
if (!g[STORE_KEY]) g[STORE_KEY] = new Map<string, Bucket>();
const buckets = g[STORE_KEY]!;

export type RateLimitResult = { ok: true } | { ok: false; retryAfterMs: number; limit: number };

/**
 * Cek apakah `key` masih boleh request.
 *
 * @param key       — identifier unik (mis. `chatbot:user:42`)
 * @param limit     — maksimum request dalam window
 * @param windowMs  — panjang window (ms)
 */
export function rateLimit(key: string, limit: number, windowMs: number): RateLimitResult {
  const now = Date.now();
  const cutoff = now - windowMs;
  let bucket = buckets.get(key);
  if (!bucket) {
    bucket = { times: [] };
    buckets.set(key, bucket);
  }
  // Buang entri di luar window
  bucket.times = bucket.times.filter((t) => t > cutoff);
  if (bucket.times.length >= limit) {
    const oldest = bucket.times[0];
    const retryAfterMs = Math.max(0, oldest + windowMs - now);
    return { ok: false, retryAfterMs, limit };
  }
  bucket.times.push(now);
  return { ok: true };
}

/** Bersihkan bucket lama (panggil periodik kalau perlu). */
export function gcRateLimits(maxAgeMs: number = 60 * 60 * 1000) {
  const cutoff = Date.now() - maxAgeMs;
  for (const [k, b] of buckets.entries()) {
    if (b.times.length === 0 || b.times[b.times.length - 1] < cutoff) {
      buckets.delete(k);
    }
  }
}

File: lib/rate_limit.py
"""Sliding-window rate limiter in-memory (per-process).
Cocok untuk single-server deployment Streamlit.
"""
from __future__ import annotations

import time
from collections import defaultdict, deque

_BUCKETS: dict[str, deque[float]] = defaultdict(deque)


def rate_limit(key: str, limit: int, window_sec: float) -> tuple[bool, float]:
    """
    Cek apakah request ke `key` masih boleh dijalankan.
    Returns: (ok, retry_after_seconds)
    """
    now = time.time()
    cutoff = now - window_sec
    bucket = _BUCKETS[key]
    while bucket and bucket[0] < cutoff:
        bucket.popleft()
    if len(bucket) >= limit:
        retry = bucket[0] + window_sec - now
        return False, max(0.0, retry)
    bucket.append(now)
    return True, 0.0
